dep0/lib/01_ast/include/dep0/ast/alpha_equivalence_impl.hpp

```diff
@@ -54,114 +114,228 @@ struct alpha_equivalence_visitor
-                        // if x and y are alpha-equivalent, they will now compare equal in all other args and ret type
-                        if (x_var.name.idx > y_var.name.idx)
-                            replace(y_var, x_var, y.args.begin() + i + 1, y.args.end(), y.ret_type.get());
-                        else
-                            replace(x_var, y_var, x.args.begin() + i + 1, x.args.end(), x.ret_type.get());
-                        return std::true_type{};
-                    },
-                    [&] (typename func_arg_t<P>::term_arg_t& x_arg) -> dep0::expected<std::true_type>
-                    {
-                        auto* const y_arg = std::get_if<typename func_arg_t<P>::term_arg_t>(&y.args[i].value);
-                        if (not y_arg)
-                            return not_alpha_equivalent(i);
-                        auto ok = is_alpha_equivalent_impl(x_arg.type, y_arg->type);
-                        if (not ok)
-                        {
-                            auto not_ok = not_alpha_equivalent(i);
-                            not_ok.reasons.push_back(std::move(ok.error()));
-                            return not_ok;
-                        }
-                        return ok;
-                    });
-            if (not ok)
-                return std::move(ok.error());
+            auto err = not_alpha_equivalent(i);
+            err.reasons.push_back(std::move(eq.error()));
+            return err;
         }
-        auto const ok = is_alpha_equivalent_impl(x.ret_type.get(), y.ret_type.get());
-        if (ok)
-            return ok;
-        else
+        if (x_arg.var and y_arg.var)
         {
-            std::ostringstream err;
-            pretty_print(err << "return type `", x.ret_type.get()) << '`';
-            pretty_print(err << " is not alpha-equivalent to `", y.ret_type.get()) << '`';
-            return dep0::error_t(err.str(), {std::move(ok.error())});
+            auto& x_var = *x_arg.var;
+            auto& y_var = *y_arg.var;
+            if (x_var == y_var)
+                continue;
+            x_var = rename(x_var, x_args.begin() + i + 1, x_args.end(), x_ret_type);
+            y_var = rename(y_var, y_args.begin() + i + 1, y_args.end(), y_ret_type);
+            if (x_var == y_var)
+                // by pure luck renaming assigned the same name to both
+                continue;
+            // renaming assigns the next unused index number counting upwards;
+            // so if index of x_var is greather than that of y_var, we know that `x_var` does not occur
+            // in the renamed y, and viceversa; we can therefore safely replace x_var in y (or viceversa);
+            // if x and y are alpha-equivalent, they will now compare equal in the remaining args and in ret type
+            if (x_var.name.idx > y_var.name.idx)
+                replace(y_var, x_var, y_args.begin() + i + 1, y_args.end(), y_ret_type);
```

Comment:
If we don't rename y_var as well, error messages might look odd and even confusing